File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/configurations/docker/DockerImageConfigurationSettings.java
package com.github.alexandrecarlton.idea.settings.layout.configurations.docker;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.github.alexandrecarlton.idea.settings.layout.configurations.ConfigurationSettings;
import java.util.List;
import java.util.Optional;
import org.immutables.value.Value;

@Value.Immutable
@JsonDeserialize(as = ImmutableDockerImageConfigurationSettings.class)
public interface DockerImageConfigurationSettings extends ConfigurationSettings {

  Optional<String> imageId();

  Optional<String> containerName();

  Optional<DockerPublishToHostInterface> publishExposedPortsToTheHostInterfaces();

  Optional<DockerExecutableSettings> executable();

  Optional<List<DockerPortBinding>> bindPorts();

  // TODO: bindMounts

  Optional<List<DockerEnvironmentVariable>> environmentVariables();

  Optional<String> runOptions();

}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/configurations/docker/DockerImageConfigurationSettings.kt
package com.github.alexandrecarlton.idea.settings.layout.configurations.docker

import com.fasterxml.jackson.annotation.JsonProperty
import com.github.alexandrecarlton.idea.settings.layout.configurations.ConfigurationSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.BeforeLaunchConfigurationSettings

data class DockerImageConfigurationSettings(

    @JsonProperty("Name")
    override val name: String,

    @JsonProperty("Share through VCS")
    override val shareThroughVcs: Boolean? = null,

    @JsonProperty("Before launch")
    override val beforeLaunch: List<BeforeLaunchConfigurationSettings>? = null,

    @JsonProperty("Image ID")
    val imageId: String? = null,

    @JsonProperty("Container name")
    val containerName: String? = null,

    @JsonProperty("Publish exposed ports to the host interfaces")
    val publishExposedPortsToTheHostInterfaces: DockerPublishToHostInterface? = null,

    @JsonProperty("Executable")
    val executable: DockerExecutableSettings? = null,

    @JsonProperty("Bind ports")
    val bindPorts: List<DockerPortBinding>? = null,

    // TODO: bindMounts

    @JsonProperty("Environment variables")
    val environmentVariables: List<DockerEnvironmentVariable>? = null,

    @JsonProperty("Run options")
    val runOptions: String? = null

) : ConfigurationSettings
